### PR TITLE
PR

### PR DESCRIPTION
**Pull request by koiiujika**

### CODE REVIEW

File: src/gate/parser/bench/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 3.24.2)
```
Comment (greye) [inline src/gate/parser/bench/CMakeLists.txt]: Arbitrary restriction, remove the line.

File: src/gate/parser/bench/lexer.l [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    TOK_AND,
+    TOK_OR,
+    TOK_NAND,
+    TOK_NOR,
+    TOK_LP, //left parenthesis
+    TOK_RP, //right paranthesis
+    TOK_E,  // equal
+    TOK_ID, //var
+    TOK_COMMA,
+};
```
Comment (greye) [inline src/gate/parser/bench/lexer.l]:
```suggestion
#include "token.h"
```

File: src/gate/parser/bench/lexer.l [hunk 3]
```diff
@@ -0,0 +1,43 @@
+%{
+#include <stdio.h>
```
Comment (greye) [inline src/gate/parser/bench/lexer.l]: The header is not used, remove.

File: src/gate/parser/bench/parser.cpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the Utopia EDA Project, under the Apache License v2.0
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021 ISP RAS (http://www.ispras.ru)
+//
+//===----------------------------------------------------------------------===//
+
+#include <FlexLexer.h>
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: What is used from the header?

File: src/gate/parser/bench/parser.cpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+#include "gate/model/gnet.h"
+#include "gate/model/gate.h"
+#include "gate/model/gsymbol.h"
+#include "headerFile"
+#include "tokens.h"
+
+#include <iostream>
+#include <stdlib.h>
+#include <vector>
+#include <map>
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: https://llvm.org/docs/CodingStandards.html#include-style

File: src/gate/parser/bench/parser.cpp [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+///Defines the compare method for TokenMap.name.
+#define COMPARE(text) do {\
+  for (int j = 0; j < strlen(yytext); j++) {\
+    if (i.name[j] == yytext[j])\
+      k += 1;\
+  }\
+  if (k == strlen(yytext)) {\
+    std::cout << "ERROR IN " << text << " WITH "\
+      << yytext << "\tline: " << line << std::endl;\
+    exit(EXIT_FAILURE);\
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Why `exit` the program?

File: src/gate/parser/bench/parser.cpp [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+  }\
+} while (false)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Use `strcmp` or move to `std::string`

File: src/gate/parser/bench/parser.cpp [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+#include <iostream>
+#include <stdlib.h>
+#include <vector>
+#include <map>
+
+extern "C" int scan_token();
+
+///Defines the compare method for TokenMap.name.
+#define COMPARE(text) do {\
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Whe the macro? Use functions instead when possible.

File: src/gate/parser/bench/parser.cpp [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    exit(EXIT_FAILURE);\
+  }\
+} while (false)
+
+///Defines the alert method for debugging information.
+#define ALERT(errorString) do {\
+  std::cout << " expected '" << errorString << "'\tline: "\
+    << line << "\tСaught " << yytext << std::endl;\
+  exit(EXIT_FAILURE);\
+} while (false)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Why the macro? Use functions instead when possible.

File: src/gate/parser/bench/parser.cpp [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    ALERT(errorString);\
+  }\
+} while (false)
+
+///Defines the adding in a token map vector.
+#define MAPS(definite, typeInit) do {\
+  char* text = new char[strlen(yytext)];\
+  strcpy(text, yytext);\
+  TokenMap map {text, definite, typeInit};\
+  maps.push_back(map);\
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: - memory allocated and never deleted
- not enough memory allocated: C-strings are 0-terminated, therefore 1 more character required

File: src/gate/parser/bench/parser.cpp [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+  int definite; // 0 = NOT init, 1 = init
+  Tokens typeInit; // 0 = input, 1 = output, 2 = function
+  int usedInGate = 0;
+  // Gate::Id gateId;
+};
+
+std::vector<TokenMap> maps;
+static std::size_t place = 0;
+static std::size_t line = 0;
+GNet net;
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Global variables are forbidden

File: src/gate/parser/bench/parser.cpp [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+///Defines the method for verifying the next token.
+#define ASSERT_NEXT_TOKEN(expectedToken, errorString) do {\
+  Tokens token = getNextToken();\
+  if ( token != expectedToken ) {\
+    ALERT(errorString);\
+  }\
+} while (false)
+
+///Defines the adding in a token map vector.
+#define MAPS(definite, typeInit) do {\
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Why the macro? Use functions instead when possible.

File: src/gate/parser/bench/parser.cpp [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+//     TOK_INPUT = 1,
+//     TOK_OUTPUT, // 2
+//     TOK_DFF, // 3
+//     TOK_NOT, // 4
+//     TOK_AND, // 5
+//     TOK_OR, // 6
+//     TOK_NAND, // 7
+//     TOK_NOR, // 8
+//   Gate::Id gateId;
+// };
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Remove

File: src/gate/parser/bench/parser.cpp [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+    if (size(mapsUpd) != size(maps)) {
+      std::cout << "ERROR IN UNKNOWN DEFINITION " << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+void input() {
+  for (auto i : maps) {
+    int k = 0;
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Unused variable?

File: src/gate/parser/bench/parser.cpp [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+      if (i.definite == 0 && j.definite == 1 &&
+            static_cast<std::string>(i.name) == static_cast<std::string>(j.name)) {
+          i.definite = 1;
+          mapsUpd.push_back(i);
+      }
+    }
+  }
+  for (auto i : maps) {
+    if(i.definite == 1)
+    mapsUpd.push_back(i);
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Broken indentation and spacing.

File: src/gate/parser/bench/parser.cpp [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+GNet net;
+
+void unknown() {
+  std::vector<TokenMap> mapsUpd;
+  int k = 0;
+  for (auto i : maps) {
+    for (auto j : maps) {
+      k += 1;
+      if (i.definite == 0 && j.definite == 1 &&
+            static_cast<std::string>(i.name) == static_cast<std::string>(j.name)) {
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Why not `strcmp`? Or make use of `std::string` consistently.

File: src/gate/parser/bench/parser.cpp [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+    if (strlen(yytext) == strlen(i.name) && i.typeInit == 1)
+      COMPARE("ENTRY INPUT");
+  }
+} 
+
+void output(Tokens type) {
+  for (auto i : maps) {
+    if (type != TOK_OUTPUT) {
+      int k = 0;
+      if (strlen(yytext) == strlen(i.name) && i.typeInit == 2)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: What is `1` and `2`?

File: src/gate/parser/bench/parser.cpp [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+  for (auto i : maps) {
+    if (strlen(yytext) == i.name.size() && i.typeInit == 1)
+      COMPARE("ENTRY INPUT", line);
+  }
+} 
+
+void output(Tokens type, std::size_t &line, std::vector<TokenMap>& maps) {
+  for (auto i : maps) {
+    if (type != TOK_OUTPUT) {
+      if (strlen(yytext) == i.name.size() && i.typeInit == 2)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: What is `1` and `2`?

File: src/gate/parser/bench/parser.cpp [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+#include <map>
+#include <memory>
+#include <stdlib.h>
+#include <vector>
+
+#include "gate/model/gate.h"
+#include "gate/model/gnet.h"
+#include "gate/model/gsymbol.h"
+#include "headerFile"
+#include "tokens.h"
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: https://llvm.org/docs/CodingStandards.html#include-style

File: src/gate/parser/bench/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+find_package(FLEX)
+FLEX_TARGET(BenchLexer 
+  lexer.l 
+  ${CMAKE_CURRENT_BINARY_DIR}/lexer.c 
+  DEFINES_FILE "${CMAKE_CURRENT_BINARY_DIR}/headerFile"
```
Comment (greye) [inline src/gate/parser/bench/CMakeLists.txt]: Project code style requires file names to be in `snake_case`.

File: src/gate/parser/bench/parser.cpp [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+extern "C" int scan_token();
+
+///Defines the compare method for TokenMap.name.
+#define COMPARE(text, line) do {\
+  if (yytext == i.name) {\
+    std::cout << "ERROR IN " << static_cast<const char*>(text) << " WITH "\
+      << yytext << "\tline: " << line << std::endl;\
+    exit(EXIT_FAILURE);\
+  }\
+} while (false)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Why the macro? Use functions instead when possible.

File: src/gate/parser/bench/parser.cpp [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+
+///Defines the alert method for debugging information.
+#define ALERT(errorString, line) do {\
+  std::cout << " expected '" << errorString << "'\tline: "\
+    << line << "\tСaught " << yytext << std::endl;\
+  exit(EXIT_FAILURE);\
+} while (false)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Why the macro? Use functions instead when possible.

File: src/gate/parser/bench/parser.cpp [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+  if ( token != expectedToken ) {\
+    ALERT(errorString, line);\
+  }\
+} while (false)
+
+///Defines the adding in a token map vector.
+#define MAPS(definite, typeInit, maps) do {\
+  char* text = new char[strlen(yytext)];\
+  strncpy(text, yytext, strlen(yytext) + 1);\
+  TokenMap map {text, definite, typeInit};\
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: - not enough memory allocated: C-strings are zero-terminated and requires an additional character
- manual allocation does not make sense after the move to `std::string`
```suggestion
  TokenMap map {yytext, definite, typeInit};\
```

File: src/gate/parser/bench/parser.cpp [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+  for (auto i : maps) {
+    if(i.definite == 1)
+    mapsUpd.push_back(i);
+  }
+  for (auto i : mapsUpd) {
+    if (size(mapsUpd) != size(maps)) {
+      std::cout << "ERROR IN UNKNOWN DEFINITION " << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: So, why it is in the loop?
```suggestion
  if (size(mapsUpd) != size(maps)) {
    std::cout << "ERROR IN UNKNOWN DEFINITION " << std::endl;
    exit(EXIT_FAILURE);
  }
```

File: src/gate/parser/bench/parser.cpp [hunk 25]
```diff
@@ -0,0 +1,290 @@
...
+  for (auto i : maps) {
+    if (i.definite == 1) { // definite = {0 if not declared, 1 if declared}.
+      if (strlen(yytext) == i.name.size() && whereCalled != TOK_OUTPUT
+        && i.typeInit != 2)  {
+        COMPARE("DOUBLE DEFINITION", line);
+        } else if (strlen(yytext) == i.name.size() 
+          && whereCalled == TOK_OUTPUT ) {
+          if(i.typeInit == 2)
+            COMPARE("DOUBLE DEFINITION", line);
+        }
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Broken indentation and spacing

File: src/gate/parser/bench/parser.cpp [hunk 26]
```diff
@@ -0,0 +1,290 @@
...
+void parseID(std::size_t &place, std::size_t &line, std::vector<TokenMap>& maps) {
+  MAPS(1, TOK_E, maps); // adding.
+  ASSERT_NEXT_TOKEN(TOK_E, "E", place, line);
+  Tokens token = getNextToken(place, line);
+  if (token == TOK_AND || token == TOK_OR || token == TOK_NAND || token == TOK_NOR) {
+    maps.back().typeInit = token;
+    parseParenthesisID(token, place, line, maps);
+  } else if (token == TOK_DFF || token == TOK_NOT) { 
+    maps.back().typeInit = token;
+    parseParenthesisIO(token, place, line, maps);
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Are branches identical?

File: src/gate/parser/bench/parser.cpp [hunk 27]
```diff
@@ -0,0 +1,290 @@
...
+  std::unique_ptr<GNet> net = std::make_unique<GNet>();
+  int dffFlag = 0;
+  Gate::Id dffClock = -1;
+
+  std::map<std::string, Gate::Id> gmap;
+  for (auto i : maps) {
+    if (i.usedInGate == 0 && gmap.find(i.name) == gmap.end()) {
+      if (i.typeInit == 1) {
+        gmap.insert(std::make_pair(i.name, net->addIn()));
+      } else if (dffFlag == 0 && i.typeInit == 3) {
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: What is `1' and `3`?

File: src/gate/parser/bench/parser.cpp [hunk 28]
```diff
@@ -0,0 +1,290 @@
...
+      } else if (_type == 6) {
+        net->setGate(arg, GateSymbol::OR, ids);
+      } else if (_type == 7) {
+        net->setGate(arg, GateSymbol::NAND, ids);
+      } else if (_type == 8) {
+        net->setGate(arg, GateSymbol::NOR, ids);
+      } else if (_type == 3) {
+        ids.push_back(Signal::always(dffClock));
+        net->setGate(arg, GateSymbol::DFF, ids );
+      }
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Use `switch` instead and `enum` instead of arbitrary literals.

File: src/gate/parser/bench/parser.cpp [hunk 29]
```diff
@@ -0,0 +1,290 @@
...
+      } 
+    }
+  }
+
+  for (auto i : maps) {
+    if (i.typeInit == 2) {
+      net->addOut(gmap[i.name]);
+    }
+  }
+  // std::cout << *net;
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Remove

**Review by greye — CHANGES_REQUESTED** [review]
- use `enum` instead of literals when appropriate
- remove commented code
- replace macros with functions when possible

File: src/gate/parser/bench/parser.cpp (hunk 26 above)
Comment (koiiujika) [reply src/gate/parser/bench/parser.cpp]: No, difference in the function names "parseParenthesisI + D" and "parseParenthesisI + O". I will rename them then.

File: src/gate/parser/bench/parser.cpp [hunk 30]
```diff
@@ -0,0 +1,376 @@
...
+  }\
+} while (false)
+
+///Defines the adding in a token map vector.
+#define MAPS(definite, typeInit, maps) do {\
+  char* text = new char[strlen(yytext)];\
+  strncpy(text, yytext, strlen(yytext) + 1);\
+  TokenMap map {text, definite, typeInit};\
+  maps.push_back(map);\
+} while (false)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Replace macros with functions.

File: src/gate/parser/bench/parser.cpp [hunk 31]
```diff
@@ -0,0 +1,376 @@
...
+using Signal = eda::gate::model::Gate::Signal;
+using Gate = eda::gate::model::Gate;
+using GateSymbol = eda::gate::model::GateSymbol;
+
+struct TokenMap {
+  std::string name;
+  int definite; // 0 = NOT init, 1 = init
+  Tokens typeInit; // 0 = input, 1 = output, 2 = function
+  int usedInGate = 0;
+  // Gate::Id gateId;
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Remove the line

File: src/gate/parser/bench/parser.cpp [hunk 32]
```diff
@@ -0,0 +1,376 @@
...
+                      std::vector<TokenMap>& maps,
+                      std::string& errorString,
+                      std::string &expectedString) {
+  for (auto i : maps) {
+    if (i.definite == 1) { // definite = {0 if not declared, 1 if declared}.
+      if (strlen(yytext) == i.name.size() && whereCalled != TOK_OUTPUT
+          && i.typeInit != TOK_OUTPUT)  {
+        COMPARE("DOUBLE DEFINITION", line, errorString, expectedString);
+      } else if (strlen(yytext) == i.name.size() 
+                 && whereCalled == TOK_OUTPUT ) {
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: It actually fits 80 characters, fyi

File: src/gate/parser/bench/parser.cpp [hunk 33]
```diff
@@ -0,0 +1,376 @@
...
+          if(i.typeInit == TOK_OUTPUT)
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Broken spacing

File: src/gate/parser/bench/parser.cpp [hunk 34]
```diff
@@ -0,0 +1,376 @@
...
+        case TOK_INPUT:
+        case TOK_OUTPUT:
+        case TOK_LP:
+        case TOK_RP:
+        case TOK_E:
+        case TOK_ID:
+        case TOK_COMMA:
+        case TOK_N:
+        case TOK_COMMENT:
+          break;
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]:
```suggestion
        default:
          break;
```

File: src/gate/parser/bench/parser.cpp [hunk 35]
```diff
@@ -0,0 +1,376 @@
...
+      }
```
Comment (greye) [inline src/gate/parser/bench/parser.cpp]: Broken indentation: `case` should have the same indentation as `switch`, e.g.
```
switch (c) {
case A:
case B:
...
}
```

**Review by greye — CHANGES_REQUESTED** [review]

**Pull request merged**